output/clickhouse.go

```diff
@@ -133,24 +136,29 @@ func shouldReconnect(err error) bool {
 }
 
 // LoopWrite will dead loop to write the records
-func (c *ClickHouse) loopWrite(batch model.Batch, callback func(batch model.Batch) error) {
+func (c *ClickHouse) loopWrite(batch *model.Batch, callback func(batch *model.Batch) error) {
 	var err error
 	times := c.chCfg.RetryTimes
 	defer statistics.FlushBatchBacklog.WithLabelValues(c.taskCfg.Name).Dec()
 	for {
 		if err = c.write(batch); err == nil {
-			callback(batch)
+			if err = callback(batch); err != nil {
+				log.Criticalf("%s: committing offset to failed with error %+v", c.taskCfg.Name, err)
+				os.Exit(-1)
+			}
 			return
 		}
 		if std_errors.Is(err, context.Canceled) {
-			log.Infof("ClickHouse.loopWrite quit due to the context has been cancelled")
+			log.Infof("%s: ClickHouse.loopWrite quit due to the context has been cancelled", c.taskCfg.Name)
 			return
 		}
 		times--
-		log.Errorf("flush batch(try #%d) failed with error %+v", c.chCfg.RetryTimes-times, err)
-		if times > 0 {
-			statistics.FlushMsgsErrorTotal.WithLabelValues(c.taskCfg.Name).Add(float64(batch.RealSize))
-			return
+		statistics.FlushMsgsErrorTotal.WithLabelValues(c.taskCfg.Name).Add(float64(batch.RealSize))
+		if times <= 0 {
+			log.Criticalf("%s: flush batch(try #%d) failed with error %+v", c.taskCfg.Name, c.chCfg.RetryTimes-times, err)
+			os.Exit(-1)
```

Comment:
here `os.Exit` may affect other tasks. It's dangerous

Reply:
Do you want nice quit of a task in this case?